File: ds_algos_primer/java/ArraysAndStrings.java
/*
 *   Title: ArraysAndStrings
 *
 *   This file contains the solutions for the Arrays and Strings exercises in
 *   the DS & Algos Primer. If you have not already attempted these exercises,
 *   we highly recommend you complete them before reviewing the solutions here.
 *
 *   Execution: javac ArraysAndStrings.java && java -ea ArraysAndStrings
 */

import java.util.*;

public class ArraysAndStrings {

    // Exercise 1.1: Write a function that takes an integer array and reverses
    // the values in place
    public static void reverseArray(int[] arr) {
        // We will iterate to the midpoint of the array. For each value, we can
        // get the index its supposed to swap with by computing arr.length-i-1
        for (int i = 0; i < arr.length/2; i++) {
            int temp = arr[i];
            arr[i] = arr[arr.length-i-1];
            arr[arr.length-i-1] = temp;
        }
    }

    // Exercise 1.2: Given a 2D matrix, write a function to print the values in
    // the matrix in a clockwise spiral from outside to inside
    public static void printSpiral(int[][] arr) {
        // We need to keep track of the boundaries of the current layer of the
        // spiral that we are traversing
        int minRow = 0;
        int minCol = 0;
        int maxRow = arr.length-1;
        int maxCol = arr[0].length-1;

        // Once the mins and maxes converge, we are at the center of the spiral.
        // The spiral follows a fixed set of steps. We go left, down, right, up.
        // For each of these, we just interate to the bounds, so we express each
        // one explicitly.
        while (minRow < maxRow && minCol < maxCol) {
            // Go across the top
            for (int col = minCol; col <= maxCol; col++) {
                System.out.println(arr[minRow][col]);
            }
            minRow++;

            // Go down the right side
            for (int row = minRow ; row <= maxRow; row++) {
                System.out.println(arr[row][maxCol]);
            }
            maxCol--;

            // Go across the bottom
            for (int col = maxCol; col >= minCol; col--) {
                System.out.println(arr[maxRow][col]);
            }
            maxRow--;

            // Go up the left side
            for (int row = maxRow; row >= minRow; row--) {
                System.out.println(arr[row][minCol]);
            }
            minCol++;
        }
    }

    // Exercise 1.3: Given a 2D matrix, write a function to print the values in
    // the matrix in a zig-zag order
    public static void printZigZag(int[][] arr) {
        int row = 0;
        int col = 0;

        // Like the spiral, we have clearly defined directions we need to go. In
        // this case we either go up to the right or down to the left. We define
        // each of these explicitly and just go back and forth between doing one
        // and the other
        while (true) {

            // Go up to the right
            while (row > 0 && col < arr[0].length-1) {
                System.out.println(arr[row][col]);
                row--;
                col++;
            }
            // Without this we won't print the final value in the diagonal
            System.out.println(arr[row][col]);

            // Check whether we're at the bottom right corner
            if (row == arr.length-1 && col == arr[0].length-1) break;

            // We need to update our positiion differently depending on whether
            // we're still going along the top of the matrix or down the
            // righthand side
            else if (col+1 < arr[0].length) col++;
            else row++;

            // Go down to the left
            while (row < arr.length-1 && col > 0) {
                System.out.println(arr[row][col]);
                row++;
                col--;
            }
            // Without this we won't print the final value in the diagonal
            System.out.println(arr[row][col]);

            // Check whether we're at the bottom right corner
            if (row == arr.length-1 && col == arr[0].length-1) break;

            // Are we going along the lefthand side or the bottom?
            else if (row+1 < arr.length) row++;
            else col++;
        }
    }

    // Sample test cases
    public static void main(String[] args) {
        int[] toReverse = new int[]{1,2,3,4,5};
        reverseArray(toReverse);
        System.out.println(Arrays.toString(toReverse));

        int[][] matrix = new int[4][5];
        int val = 1;
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[0].length; j++) {
                matrix[i][j] = val++;
            }
        }
        printSpiral(matrix);

        printZigZag(matrix);
    }
}

File: ds_algos_primer/python/arrays_and_strings.py
"""
Title: Arrays and Strings

This file contains the template for the Arrays and Strings exercises in
the DS & Algos Primer. Fill in the exercises here and refer to
arrays_and_strings_solutions.py for the complete code samples.

Execution: python arrays_and_strings.py


*** IMPORTANT NOTE ***
Python provides a lot of inbuilt functions to accomplish certain tasks. If you
are aware of these, that's great.

HOWEVER, the goal of these exercises is to understand these data structures.
Therefore, you are discouraged from writing one- to two-line functions. Instead
you will learn a lot more by implementing these things manually.

In your interview, you may want to use these inbuilt functions, but while
learning, you will learn more by doing things the hard way.
"""

"""
Exercise 1.1: Write a function that takes an integer array and reverses
the values in place
"""
def reverse_array(arr: List[int]):
    # INSERT YOUR SOLUTION HERE

"""
Exercise 1.2: Given a 2D matrix, write a function to print the values in
the matrix in a clockwise spiral from outside to inside
"""
def print_spiral(arr: List[List[int]]):
    # INSERT YOUR SOLUTION HERE

"""
Exercise 1.3: Given a 2D matrix, write a function to print the values in the
matrix in a zig-zag order
"""
def print_diagonals(arr: List[List[int]]):
    # INSERT YOUR SOLUTION HERE

"""
Exercise 1.4: Write a function that takes in a string and removes every
even-indexed character
"""
def remove_even(s: str) -> str:
    # INSERT YOUR SOLUTION HERE

"""
Exercises 1.5: Zig Zag Conversion
Full Problem Definition: https://leetcode.com/problems/zigzag-conversion/
"""
def zig_zag(s: str, num_rows: int) -> str:
    # INSERT YOUR SOLUTION HERE

"""
Exercise 2.1: Given a string, print out all of the substrings
"""
def print_substrings(s: str):
    # INSERT YOUR SOLUTION HERE

"""
Exercise 2.2: Write a function to find all duplicates in an array. The array
will contain exactly 1 duplicated value
"""
def find_duplicates(arr: List[int]) -> int:
    # INSERT YOUR SOLUTION HERE

"""
Exercise 2.3: Given a sorted array, find every pair of values in the
array that sum up to a given target
"""
def two_sum(arr: List[int], target: int) -> List[List[int]]:
    # INSERT YOUR SOLUTION HERE

"""
Exercise 3.1: Given two arrays, compare them to see if they are equal
"""
def arrays_are_equal(arr1: List[int], arr2: List[int]) -> bool:
    # INSERT YOUR SOLUTION HERE

"""
Exercise 3.2: Given two strings, determine if one string is the reverse of the
other string
"""
def strings_are_opposite(s1: str, s2: str) -> bool:
    # INSERT YOUR SOLUTION HERE

"""
Exercise 3.3: Given two strings, determine whether they are anagrams of
each other
"""
def are_anagrams(s1: str, s2: str) -> bool:
    # INSERT YOUR SOLUTION HERE

"""
Exercise 4.1: Given an array, compute the sum of each length-k subarray
"""
def subarray_sums(arr: List[int], k: int) -> List[int]:
    # INSERT YOUR SOLUTION HERE

"""
Exercise 4.2: Given a string, find the longest substring of the string that does
not contain any repeated characters
"""
def no_repeated_chars(s: str) -> int:
    # INSERT YOUR SOLUTION HERE

"""
Exercise 4.3: Given two strings, s and p, find all occurrences of anagrams of p
in s. The output is the starting index of each anagram
"""
def find_all_anagrams(s: str, p: str) -> List[int]:
    # INSERT YOUR SOLUTION HERE

"""
Exercise 4.4: Given two strings, s and p, find the smallest substring of s that
contains all the characters in p
"""
def smallest_substring(s: str, p: str) -> str:
    # INSERT YOUR SOLUTION HERE



# Sample test cases
if __name__ == '__main__':
    # INSERT YOUR TESTS HERE
